bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -430,6 +444,148 @@ sequentialIterations output !initialSize !batchSize !batchCount !seed !tbl =
   where
     g0 = initGen initialSize batchSize batchCount seed
 
+-------------------------------------------------------------------------------
+-- pipelined
+-------------------------------------------------------------------------------
+
+{- One iteration of the protocol for one thread looks like this:
+
+1. Lookups (db_n-1) tx_n+0
+2. Sync ?  (db_n+0, updates)
+3. db_n+1 <- Dup (db_n+0)
+   Updates (db_n+1) tx_n+0
+4. Sync !  (db_n+1, updates)
+
+Thus for two threads running iterations concurrently, it looks like this:
+
+1. Lookups (db_n-1) tx_n+0        3. db_n+0 <- Dup (db_n-1)
+                                     Updates (db_n+0) tx_n-1
+2. Sync ?  (db_n+0, updates)  <-  4. Sync !  (db_n+0, updates)
+3. db_n+1 <- Dup (db_n+0)         1. Lookups (db_n+0) tx_n+1
+   Updates (db_n+1) tx_n+0
+4. Sync !  (db_n+1, updates)  ->  2. Sync ?  (db_n+1, updates)
+1. Lookups (db_n+1) tx_n+2        3. db_n+2 <- Dup (db_n+1)
+                                     Updates (db_n+2) tx_n+1
+2. Sync ?  (db_n+2, updates)  <-  4. Sync !  (db_n+2, updates)
+3. db_n+3 <- Dup (db_n+2)         1. Lookups (db_n+2) tx_n+3
+   Updates (db_n+3) tx_n+2
+4. Sync !  (db_n+3, updates)  ->  2. Sync ?  (db_n+3, updates)
+1. Lookups (db_n+3) tx_n+4        3. db_n+4 <- Dup (db_n+3)
+                                     Updates (db_n+4) tx_n+3
+2. Sync ?  (db_n+4, updates)  <-  4. Sync !  (db_n+4, updates)
+3. db_n+5 <- Dup (db_n+4)         1. Lookups (db_n+4) tx_n+5
+   Updates (db_n+5) tx_n+4
+4. Sync !  (db_n+5, updates)  ->  2. Sync ?  (db_n+5, updates)
+
+And the initial setup looks like this:
+
+   db_1 <- Dup (db_0)
+   Lookups (db_0) tx_0
+   Updates (db_1, tx_0)
+   Sync !  (db_1, updates)    ->
+                                  1. Lookups (db_0) tx_1
+
+                                  2. Sync ?  (db_1, updates)
+1. Lookups (db_1) tx_2            3. db_2 <- Dup (db_1)
+                                     Updates (db_2) tx_1
+2. Sync ?  (db_2, updates)    <-  4. Sync !  (db_2, updates)
+3. db_3 <- Dup (db_2)             1. Lookups (db_2) tx_3
+   Updates (db_3) tx_2
+4. Sync !  (db_3, updates)        2. Sync ?  (db_3, updates)
+-}
```

Comment:
I assume `tx` is short for "transaction", but in the Cardano sense it's not a single transaction, but rather a block consisting of multiple transactions. Is it a transaction in the sense of a database transaction?

Reply:
It's really a batch of operations in the benchmark workload. It'd correspond to a Cardano block.